src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1185,38 +1119,19 @@ public ImmutableArray<DocumentId> GetDocumentIdsWithFilePath(string? filePath)
         if (string.IsNullOrEmpty(filePath))
             return [];
 
-        if (!_filePathToDocumentIdsMap.TryGetValue(filePath, out var documentIds))
-            return [];
-
-        // If this wasn't the result of a freeze, then we can return the document ids as is.  They should be accurate.
-        if (!_filePathToDocumentIdsMap.IsFrozen)
-        {
-            Debug.Assert(documentIds.All(ContainsAnyDocument));
-            return documentIds;
-        }
-
-        // We were frozen.  So we may be seeing document ids that no longer exist within this snapshot.  If so,
-        // filter them out.
-        using var _ = ArrayBuilder<DocumentId>.GetInstance(documentIds.Length, out var result);
+        return ImmutableInterlocked.GetOrAdd(
+            ref _lazyFilePathToRelatedDocumentIds,
```

Comment:
>_lazyFilePathToRelatedDocumentIds [](http://example.com/codeflow?start=16&length=33)

would be interesting to get some statistics for load/edit scenarios about how much of this dictionary gets populated. I like the simplicity in this PR, but I'm a little concerned about losing the sharing that was happening by updating/reusing FilePathToDocumentIdsMap.

Reply:
I just did an audit of hte codebase.  We only use this facility in two ways afaict:

1. at the feature level, where features want to do something special in the case of linked files.  This is almost looking at just the file the user is in.  Though occasionally it is more.  For example, as Find-Refs goes searching, and hits symbols to cascade to, it will check if those symbols are themselves in linked files.  However, in both cases, the number of files checked should be much much less than the total number of files overall.
2. at the workspace level when *mutating* a file.  In the case of mutation, when we change a file, we go and update all the linked versions of it to have the same contents.  So that's only looking at a tiny handful of files.  
3. at the workspace level when creating the full solution instance.  When doing this we effectively look at every filepath.  That's because, per file we will try to unify it with all linked files.  However, the approach i'm taking here should be good on perf for that.  That's because we process each added project one at a time, and then the added documents within that project.  When we do that, we'll force the filepathmaps to be created for the *other* project states, so we can look within them.  The sharing will still be excellent as those other project-states will not be changed, and thus will keep those maps around.  Then, once done with teh current project, we'll move to the next.  Processing *that* project will cause us to make the filepathmap for the project we just finished with.